src/Temporalio/Common/MetricCounter.cs

```diff
@@ -0,0 +1,37 @@
+using System.Collections.Generic;
+
+namespace Temporalio.Common
+{
+    /// <summary>
+    /// Metric for adding values as a counter.
+    /// </summary>
+    /// <typeparam name="T">Type of value for the metric.</typeparam>
+    public abstract class MetricCounter<T> : Metric<T>
+        where T : struct
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MetricCounter{T}" /> class.
+        /// </summary>
+        /// <param name="details">Details.</param>
+        internal MetricCounter(MetricDetails details)
+            : base(details)
+        {
+        }
+
+        /// <summary>
+        /// Add the given value to the counter.
+        /// </summary>
+        /// <param name="value">Value to add. Currently this can only be a positive integer.</param>
+        /// <param name="extraTags">Extra tags if any. If this is called multiple times with the
+        /// same tags, use <see cref="WithTags" /> for better performance.</param>
+        public abstract void Add(
+            T value, IEnumerable<KeyValuePair<string, object>>? extraTags = null);
+
+        /// <summary>
+        /// Create a new counter with the given tags.
+        /// </summary>
+        /// <param name="tags">Tags to append to existing tags.</param>
+        /// <returns>New counter.</returns>
+        public abstract MetricCounter<T> WithTags(IEnumerable<KeyValuePair<string, object>> tags);
```

Comment:
Why not have WithTags be required by `Metric<T>`?

Reply:
Because the return type is unique to each metric